torch/distributed/fsdp/_init_utils.py

```diff
@@ -0,0 +1,364 @@
+            f"ignored and is not well-supported: {module}"
+        )
+    # Include nested FSDP modules' ignored modules
+    for submodule in root_module.modules():
+        if isinstance(submodule, fsdp_file.FullyShardedDataParallel):
+            assert hasattr(submodule, "_ignored_modules")
+            ignored_modules.update(submodule._ignored_modules)
+    return ignored_modules
+
+
+def _get_ignored_params(
+    root_module: torch.nn.Module,
+    ignored_modules: Set[torch.nn.Module],
+) -> Tuple[Set[torch.nn.Parameter], Set[str]]:
+    """
+    Returns the parameters of the modules in ``ignored_modules``,
+    excluding any :class:`FlatParameter` s, and their fully prefixed names,
+    both as :class:`set` s.
+    """
+    ignored_params = set(
+        p for m in ignored_modules for p in m.parameters() if not _is_fsdp_flattened(p)
+    )
+    # Conservatively include all shared parameters' names
+    param_to_unflat_param_names = _get_param_to_unflat_param_names(
+        root_module,
+        dedup_shared_params=False,
+    )
+    ignored_param_names = set()
+    for param in ignored_params:
+        unflat_param_names = param_to_unflat_param_names[param]
+        clean_names = []
+        for k in unflat_param_names:
+            # Clean any module wrapper prefixes in case of nested wrapping
+            clean_names.append(clean_tensor_name(k))
+        ignored_param_names.update(clean_names)
+    return ignored_params, ignored_param_names
+
+
+def _get_buffer_names(root_module: nn.Module) -> Set[str]:
+    """
+    Returns the fully prefixed names of all buffers in the module hierarchy
+    rooted at ``root_module`` as a class:`set`.
+    """
+
+    def module_fn(module: nn.Module, prefix: str, buffer_names: Set[str]):
+        for buffer_name, _ in module.named_buffers(recurse=False):
+            # Clean module wrapper prefixes in case of nested wrapping
+            prefixed_buffer_name = clean_tensor_name(prefix + buffer_name)
+            buffer_names.add(prefixed_buffer_name)
+
+    def return_fn(buffer_names: Set[str], *args):
+        return buffer_names
+
+    buffer_names: Set[str] = set()
+    return _apply_to_modules(
+        root_module,
+        module_fn,
+        return_fn,
+        buffer_names,
+    )
+
+
+def _check_single_device_module(
+    module: nn.Module,
+    ignored_params: Set[nn.Parameter],
+) -> None:
+    """
+    Raises an error if ``module`` has original parameters on multiple devices,
+    ignoring the parameters in ``ignored_params``. Thus, after this method, the
+    module must be either fully on the CPU or fully on a non-CPU device.
+    """
+    devices = set(param.device for param in _get_orig_params(module, ignored_params))
+    if len(devices) > 1:
+        raise RuntimeError(
+            f"FSDP only supports single device modules but got params on {devices}"
+        )
+
+
+def _get_device_from_device_id(
+    device_id: Optional[Union[int, torch.device]],
```

Comment:
looks like the implementation also supports str

Reply:
The `device_id` argument to the FSDP constructor is currently documented to only support `int` or `torch.device`. We should follow up to see if we want to allow passing `device_id="cuda"` or not.